dir-index.cgi

```diff
@@ -197,48 +110,129 @@ sub printh1 {
+  print "<div class='container'>\n";
+  print "<h1>Page not found</h1>\n";
   print "<hr>";
-  print "<p>The requested resource could not be found on the server.</p>";
+  printf "<p>The requested resource <span class='sh'>%s</span> could not be found on the server.</p>", htmlenc($virt);;
   print "<p>Try returning to the <a href=\"/\">root directory</a> to browse available directories.</p>";
-  print "</body></html>";
+  print "</div>\n";
+  print "</body></html>\n";
 }
 
-sub printheader {
+# printtop() - Print the stuff at the "top of the page"
+#  - HTTP headers (only if being run as a CGI - but see comment below)
+#  - HTML content including
+#     - <!DOCTYPE html>
+#     - opening <html> tag
+#     - <head> including title and style sheets
+#     - opening <body> tag
+#
+sub printtop {
   my $virt = shift;
+  my $htmlstatus = shift;
+
+  # HTTP Headers, followed by \n\n
+
+  use constant IS_CGI => exists $ENV{'GATEWAY_INTERFACE'};
+  # Note - to avoid displaying the HTTP headers in the browser during testing,
+  #    wrap the next few lines in `if (IS_CGI) { ... }`
+  # Or don't worry about it. It's sensible to ignore the HTTP status lines during testing
+  #    since that leaves one less bit of machinery to go wrong in production...
+  if ($htmlstatus ne "") {
+    printf "%s\n", $htmlstatus;
+  }
   print "Content-type:text/html\n\n";
-  print "<!-- This directory index page generated on the fly by dir-index.cgi -->\n";
-  print "<html><head>\n";
-  print $stylecss;
-  printf "<title>Index of %s</title></head>\n<body>\n", htmlenc($virt);
-  printh1($virt);
-  print "<hr>";
+
+  # HTML heading
+  print "<!DOCTYPE html>\n";
+  print "<html lang='en'>\n";
+  print "<head>\n";
+  print "<meta charset='utf-8'/>\n";
+  print "<link rel='stylesheet' href='./static/style.css' />\n";
+  printf "<title>Index of %s</title>\n", $virt;
+  print "</head>\n";
+
+  # HTML body
+  print "<body>\n";
+  print "<!-- This directory index page is auto-generated by dir-index.cgi -->\n";
+  print "<!-- Do not edit this file -->\n";
+
+  printheader();
+}
+
+# printjson
+#
+sub printjson {
+  my $phys = shift;
+  my $query_string = shift;
+
+  my ($type, $mindepth, $maxdepth) = ('f', 0, 3);
+
+  if ($query_string =~ m!\btargets!) {
+    ($type, $mindepth, $maxdepth) = ('d', 2, 2);
+  }
+
+  if (open F, '-|', 'find', $phys, '-mindepth', $mindepth, '-maxdepth', $maxdepth, '-type', $type) {
+    my @list;
+
+    while (defined(my $line = readline F)) {
+      chomp $line;
+      push @list, substr $line, length $phys;
+    }
+    close F;
+
+    # output the HTTP Content-Type header and a blank line
+    print "Content-Type: application/json\n\n";
+
+    # output the JSON of the file
+    print encode_json (\@list);
+  }
+
+}
+
+# printheader - print the banner at the top of the page
+#
+sub printheader{
+    print <<EOT;
+      <header>
+      <div>
+        <img src="./static/logo.png" alt="Download OpenWrt Firmware">
```

Comment:
Maybe we should consider embedding the logo as SVG directly into the CSS file. One less resource file to worry about.